### ISSUE TITLE
Question about the label of detail prediction

### ISSUE THREAD

**Issue by HeaderZzj** [ZHKKKe/MODNet#48]
I want to know about how you get the label of detail branch.
Due to figure.2 in your paper, my code is as follow:
    image = cv2.imread("/mnt/SegmentData/labels/00000.jpg")
    kernel = np.ones((80, 80), np.uint8)
    dil = cv2.dilate(image, kernel, iterations=1)
    ero = cv2.erode(image, kernel, iterations=1))
    dil_ero = dil - ero
    out = dil_ero * image
But cannot get the similar result as you show as details_dp in Figure2.
It would be helpful if you could correct my codes. Thanks

**Comment by ZHKKKe** [comment]
Hi, thanks for your attention.

In your code, `out` is the ground truth of the detail branch, right?
During the training, you need to constrain the detail loss inside the regions of `dil_ero` (the pixel value in `dil_ero` can be 1 or 0).
your detail loss may be `detail_loss = dil_ero * l1_loss(image - pred_detail)`, where `pred_detail` includes the predicted details.
And, could you visualize `dil_ero` and `out` for me?

**Comment by HeaderZzj** [comment]
> Hi, thanks for your attention.
> 
> In your code, `out` is the ground truth of the detail branch, right?
> During the training, you need to constrain the detail loss inside the regions of `dil_ero` (the pixel value in `dil_ero` can be 1 or 0).
> your detail loss may be `detail_loss = dil_ero * l1_loss(image - pred_detail)`, where `pred_detail` includes the predicted details.
> And, could you visualize `dil_ero` and `out` for me?
[](url)
Thanks for your reply !
I misunderstood the loss and prediction of the detail branch, my bad.
My results of dil_ero could be different due to different size of kernel.
kernel size 5*5 :
![dil_ero](https://user-images.githubusercontent.com/32046786/104393495-8fff1a80-557f-11eb-8cc3-42190a43abb1.png)

kernel size 80*80: 
![dil_ero](https://user-images.githubusercontent.com/32046786/104393454-76f66980-557f-11eb-8d96-8ae25b4df0e7.png)

**Comment by HeaderZzj** [comment]
> Hi, thanks for your attention.
> 
> In your code, `out` is the ground truth of the detail branch, right?
> During the training, you need to constrain the detail loss inside the regions of `dil_ero` (the pixel value in `dil_ero` can be 1 or 0).
> your detail loss may be `detail_loss = dil_ero * l1_loss(image - pred_detail)`, where `pred_detail` includes the predicted details.
> And, could you visualize `dil_ero` and `out` for me?



And my loss is:
loss_semantic = 1/2 * loss_l2(pred_semantic, mask1_4)
loss_detail = dil_ero * loss_l1(pred_detail, mask)
loss_matte = loss_l1(pred_matte, mask) + loss_composition(pred_matte, mask)
loss = loss_semantic + loss_detail + 10*loss_matte

I wonder whether the loss was correct?

**Comment by ZHKKKe** [comment]
It seems your loss is correct. You can try to increase the scale of `loss_semantic`, as:
```
loss = 10*loss_semantic + loss_detail + 10*loss_matte
```
For `dil_ero`, you may try `kernel_size=50` and convert it to a binary mask.

A good new is that our training code is coming soon.

**Comment by HeaderZzj** [comment]
> It seems your loss is correct. You can try to increase the scale of `loss_semantic`, as:
> 
> ```
> loss = 10*loss_semantic + loss_detail + 10*loss_matte
> ```
> 
> For `dil_ero`, you may try `kernel_size=50` and convert it to a binary mask.
> 
> A good new is that our training code is coming soon.

Appreciate all that works and replys. Thank you !

**Comment by LianShuaiLong** [comment]
> Hi, thanks for your attention.
> 
> In your code, `out` is the ground truth of the detail branch, right?
> During the training, you need to constrain the detail loss inside the regions of `dil_ero` (the pixel value in `dil_ero` can be 1 or 0).
> your detail loss may be `detail_loss = dil_ero * l1_loss(image - pred_detail)`, where `pred_detail` includes the predicted details.
> And, could you visualize `dil_ero` and `out` for me?

In your code,**dil_ero** is equal to $m_d$ in equation(3):$L_d = m_d||d_p − \alpha_g||_1$

**Comment by yarkable** [comment]
@ZHKKKe Thanks for your good work. And may I ask that when the code will be released?

**Comment by ZHKKKe** [comment]
@yarkable scheduled in next two weeks.

**Comment by yarkable** [comment]
> 
> 
> @yarkable scheduled in next two weeks.

Okay, got it.

**Comment by ZHKKKe** [comment]
The main training code is released.
